File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
"""
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None:
            return head
        
        prev = head
        cur = head.next
        head.next = None
        
        while cur.next != None:
            next_node = cur.next
            cur.next = prev
            prev = cur
            cur = next_node
            
        cur.next = prev     
        
        return cur
""" 

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None:
            return head
        
        return self.reverse(None, head)
    
    def reverse(self, prev, cur):
        if cur.next == None:
            cur.next = prev
            return cur
        
        nxt = cur.next
        cur.next = prev
        return self.reverse(cur, nxt)
        
        
            
        

File: 0206-reverse-linked-list/0206-reverse-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
/*
func reverseList(head *ListNode) *ListNode {
    var prev *ListNode
    cur := head
    
    for cur != nil {
        next := cur.Next
        cur.Next = prev
        prev = cur
        cur = next
    }
    
    return prev
}
*/

func reverse(prev, cur *ListNode) *ListNode {
    if cur == nil {
        return prev
    }
    
    next := cur.Next
    cur.Next = prev
    
    return reverse(cur, next)
}

func reverseList(head *ListNode) *ListNode {
    if head == nil {
        return head
    }
    
    return reverse(nil, head)
}